samcli/lib/build/build_strategy.py

```diff
@@ -163,6 +163,10 @@ def build_single_layer_definition(self, layer_definition: LayerBuildDefinition)
             )
 
         single_build_dir = layer.get_build_dir(self._build_dir)
+
+        # pass down a copy of env_vars, to avoid modifying the original env_vars dict
+        container_env_vars = deepcopy(layer_definition.env_vars)
```

Comment:
I wonder if we need a better way to do this? It seems like we are doing this in a couple places and maybe it is the right thing but seems like the next time we add a resource to build, we will need to do this all over again. @mndeveci what are your thoughts? I am not as familiar with the caching logic as you are.

Not opposed to this, just trying to think through if we can make this better for the future.

Reply:
That is a good point @jfuss. I missed layers in the first PR, and this one will cover that too. And the root cause is coming from pass-by-reference. This `env_vars` dictionary is altered in the way below which changes the `build_definition` signature which causes re-builds for cached build.

But instead of `deepcopy`'ing in different places, we can think about moving `env_vars` property of [`FunctionBuildDefinition`](https://github.com/aws/aws-sam-cli/blob/develop/samcli/lib/build/build_graph.py#L406) and [`LayerBuildDefinition`](https://github.com/aws/aws-sam-cli/blob/develop/samcli/lib/build/build_graph.py#L350) into [`AbstractBuildDefinition`](https://github.com/aws/aws-sam-cli/blob/develop/samcli/lib/build/build_graph.py#L320) and then add a property method which will return the copied instance of dictionary.

```
class AbstractBuildDefinition:
    """
    Abstract class for build definition
    Build definition holds information about each unique build
    """

    def __init__(self, source_md5: str, env_vars: Optional[Dict] = None) -> None:
        self.uuid = str(uuid4())
        self.source_md5 = source_md5
        self._env_vars = env_vars if env_vars else {}

    @property
    def env_vars(self):
        return deepcopy(self._env_vars)
```

@Macok really appreciate your work on this, would you be willing to update and test your implementation with the one I suggested above?

Reply:
@mndeveci @jfuss These are all valid points and I like the solution, let me fix this quickly

Reply:
@mndeveci @jfuss Done :+1: